include/alpaka/kernel/TaskKernelGenericSycl.hpp

```diff
@@ -136,32 +231,80 @@ namespace alpaka
+              supported = true;
+            } else {
+#   if(SYCL_SUBGROUP_SIZE == 0)
+              // no explicit SYCL target, assume JIT compilation
+              LAUNCH_SYCL_KERNEL_IF_SUBGROUP_SIZE_IS(sub_group_size)
+              supported = true;
 #    else
-                    auto acc = TAcc{
-                        item_elements,
-                        work_item,
-                        dyn_shared_accessor,
-                        st_shared_accessor,
-                        global_fence_dummy,
-                        local_fence_dummy};
+              // check if the kernel should be launched with a subgroup size of 4
+              if constexpr(sub_group_size == 4) {
+#    if(SYCL_SUBGROUP_SIZE & 4)
+                LAUNCH_SYCL_KERNEL_IF_SUBGROUP_SIZE_IS(4)
+#    else
+                // empty kernel, required to keep SYCL happy
+                LAUNCH_EMPTY_SYCL_KERNEL
 #    endif
-                    core::apply(
-                        [k_func, &acc](typename std::decay_t<TArgs> const&... args) { k_func(acc, args...); },
-                        k_args);
-                });
+                supported = true;
+              } 
+ 
+              // check if the kernel should be launched with a subgroup size of 8
+              if constexpr(sub_group_size == 8) {
+#    if(SYCL_SUBGROUP_SIZE & 8)
+                LAUNCH_SYCL_KERNEL_IF_SUBGROUP_SIZE_IS(8)
+#    else
+                // empty kernel, required to keep SYCL happy
+                LAUNCH_EMPTY_SYCL_KERNEL
+#    endif
+                supported = true;
+              } 
+ 
+              // check if the kernel should be launched with a subgroup size of 16
+              if constexpr(sub_group_size == 16) {
+#    if(SYCL_SUBGROUP_SIZE & 16)
+                LAUNCH_SYCL_KERNEL_IF_SUBGROUP_SIZE_IS(16)
+#    else
+                // empty kernel, required to keep SYCL happy
+                LAUNCH_EMPTY_SYCL_KERNEL
+#    endif
+                supported = true;
+              } 
+ 
+              // check if the kernel should be launched with a subgroup size of 32
+              if constexpr(sub_group_size == 32) {
+#    if(SYCL_SUBGROUP_SIZE & 32)
+                LAUNCH_SYCL_KERNEL_IF_SUBGROUP_SIZE_IS(32)
+#    else
+                // empty kernel, required to keep SYCL happy
+                LAUNCH_EMPTY_SYCL_KERNEL
+#    endif
+                supported = true;
+              } 
+ 
+              // check if the kernel should be launched with a subgroup size of 64
+              if constexpr(sub_group_size == 64) {
+#    if(SYCL_SUBGROUP_SIZE & 64)
+                LAUNCH_SYCL_KERNEL_IF_SUBGROUP_SIZE_IS(64)
+#    else
+                // empty kernel, required to keep SYCL happy
+                LAUNCH_EMPTY_SYCL_KERNEL
+#    endif
+                supported = true;
+              } 
+#    endif
+
+              // this subgroup size is not support, raise an exception
+              if (not supported)
+                throw sycl::errc::kernel_not_supported;
```

Comment:
Please only throw objects derived from `std::exception`, not an enumerator :)

Reply:
As far as I can see, this is the behaviour described by the SYCL specification when launching SYCL kernels:

> [4.9.4.2. SYCL functions for invoking kernels](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:invokingkernels)
> 
> ```c++
> template <int Dimensions>
> void parallel_for(range<Dimensions> numWorkItems, const kernel& kernelObject)
> ```
> 
> ...
> Throws an `exception` with the `errc::kernel_not_supported` error code if the `kernelObject` is not compatible with either the device associated with the primary queue of the [command group](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#command-group) or with the device associated with the secondary queue (if specified).

Currently, `alpaka::TaskKernelGenericSycl::operator()` does not catch any of the exception that are potentially thrown by `cgh.parallel_for(...)`, so the `sycl::errc` exceptions would propagate through to the user code.

So this PR is just making things more consistent :-)

Later we may decide to catch all `sycl::errc` exceptions and wrap them in something that derives from `std::exception`; if we do that, we will change this exception to match that behaviour.

Reply:
I've looked a bit into exceptions in SYCL and maybe something like `sycl::exception(sycl::make_error_code(sycl::errc::kernel_not_supported))` could work, since `sycl::exception` inherits from `std::exception`

Reply:
@fwyzard the SYCL documentation says it

> Throws an `exception` with the `errc::kernel_not_supported`.

This is different from throwing a value of `errc::kernel_not_supported` directly.

Also the documentation says ([here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_error_handling)):

> Any error reported by a SYCL backendmust derive from the base `sycl::exception`.

And `sycl::exception` derives from `std::exception`, see [here](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:exception.class).

So everything thrown out of SYCL is derived from `std::exception`.

@AuroraPerego your suggestions seems good to me.

Reply:
Thanks for the clarification!